pkg/engine/diff.go

```diff
@@ -230,7 +230,21 @@ func printObject(
 // GetResourceOutputsPropertiesString prints only those properties that either differ from the input properties or, if
 // there is an old snapshot of the resource, differ from the prior old snapshot's output properties.
 func GetResourceOutputsPropertiesString(
-	step StepEventMetadata, indent int, planning bool, debug bool) string {
+	step StepEventMetadata, indent int, planning bool, debug bool, refresh bool) string {
+	// We should only print outputs if the outputs are known to be complete. This will be the case if we are
+	//   1) not doing a preview
+	//   2) doing a refresh
+	//   3) doing a read
+	//
+	// Technically, 2 and 3 are the same, since they're both bottoming out at a provider's implementation of Read, but
+	// the upshot is that either way we're ending up with outputs that are exactly accurate. If we are not sure that we
+	// are in one of the above states, we shouldn't try to print outputs.
+	if planning && !refresh {
+		if step.Op != deploy.OpRead && step.Op != deploy.OpReadReplacement {
+			return ""
```

Comment:
note: if you're doing all &&s, i'm fine with this on one line.

Reply:
Also, this logic is definitely *not* what you had before.  Before it was:

```
	if planning && (!refresh || step.Op == deploy.OpRead || step.Op == deploy.OpReadReplacement) {
		return ""
```

Consider if you had:

planning=true
refresh=false
step.op=opread.

with the original code that would return "".  But now it won't.  

This is one of my beefs with complex expressions htat mix and match ```!```, ```&&``` and ```||```.  It's *really* easy to write something that is not correct, and it's very hard to see the flow of what was intended.  For example, with your original code, it would have been clear to see:

```go
if planning {
    if !refresh {
        return "";
    }
    if (step.Op == deploy.OpRead || step.Op == deploy.OpReadReplacement) {
        return "";
    }
}
```

This clearly breaks out the different types of questions.  First, if we're not planning, we don't care and we go ahead and do the normal stuff.  However, if we were planning, then the next question is: was this a refresh or not?  If not, then we definitely don't want to show anything.  Then, once we know it's a refresh, we don't care to show the refresh for reads.  

In pieces the flow is clear, and it's obvious what must be a precondition prior to a check as well as what the shortcircuit checks are.    With a large combined expression, that's not clear.  And either the checked in code is wrong, or the original code is wrong.  But i can't tell one way or the other :)